File: src/main/java/wraith/fabricaeexnihilo/compatibility/rei/GlyphWidget.java
package wraith.fabricaeexnihilo.compatibility.rei;

import com.mojang.blaze3d.systems.RenderSystem;
import me.shedaniel.math.Rectangle;
import me.shedaniel.rei.api.client.gui.widgets.WidgetWithBounds;
import net.minecraft.client.gui.DrawContext;
import net.minecraft.client.gui.Element;
import net.minecraft.client.util.math.MatrixStack;
import net.minecraft.util.Identifier;

import java.util.Collections;
import java.util.List;

public class GlyphWidget extends WidgetWithBounds {
    private final int width;
    private final int height;
    private final int x;
    private final int y;
    private final Rectangle bounds;
    private final Identifier texture;
    private final int u;
    private final int v;

    public GlyphWidget(Rectangle bounds, int x, int y, int width, int height, Identifier texture, int u, int v) {
        this.bounds = new Rectangle(bounds);
        this.texture = texture;
        this.u = u;
        this.v = v;
        this.x = x;
        this.y = y;
        this.width = width;
        this.height = height;
    }

    @Override
    public Rectangle getBounds() {
        return bounds;
    }

    @Override
    public void render(DrawContext context, int mouseX, int mouseY, float delta) {
        context.drawTexture(texture, x, y, u, v, width, height);
    }

    @Override
    public List<? extends Element> children() {
        return Collections.emptyList();
    }
}

File: src/main/kotlin/exnihilofabrico/compatibility/rei/GlyphWidget.kt
package exnihilofabrico.compatibility.rei

import com.mojang.blaze3d.systems.RenderSystem
import me.shedaniel.math.api.Rectangle
import me.shedaniel.rei.gui.widget.RecipeBaseWidget
import net.minecraft.client.MinecraftClient
import net.minecraft.util.Identifier

data class GlyphWidget(val rectangle: Rectangle, val x: Int, val y: Int, val width: Int, val height: Int, val texture: Identifier, val u: Int, val v: Int): RecipeBaseWidget(rectangle) {
    override fun render(mouseX: Int, mouseY: Int, delta: Float) {
        RenderSystem.disableLighting()
        MinecraftClient.getInstance().textureManager.bindTexture(texture)
        val sprite = MinecraftClient.getInstance().getSpriteAtlas(texture).apply(texture)
        RenderSystem.color4f(1.0f, 1.0f, 1.0f, 1.0f)
        blit(x, y, u, v, width, height)
    }
}
